File: src/main/java/org/afterlike/lucid/Lucid.java
package org.afterlike.lucid;

import java.util.Objects;
import org.afterlike.lucid.task.DelayedTaskHandler;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import re.tsuku.fastbus.FastBus;

public class Lucid {
	private static final Logger LOGGER = LogManager.getLogger();
	private static final Lucid INSTANCE = new Lucid();
	private static final String VERSION = org.afterlike.lucid.BuildConstants.VERSION;
	private final FastBus eventBus;
	public Lucid() {
		this.eventBus = new FastBus();
	}

	public void initialize() {
		final long startTime = System.nanoTime();
		eventBus.subscribe(DelayedTaskHandler.get());
		LOGGER.info("Initialized in {}ms.", (System.nanoTime() - startTime) / 1_000_000);
	}

	public void lateInitialize() {
		// late initialization logic
	}

	public static Lucid get() {
		return Objects.requireNonNull(INSTANCE);
	}

	public FastBus getEventBus() {
		return eventBus;
	}

	public String getVersion() {
		return VERSION;
	}
}

File: src/main/kotlin/org/afterlike/lucid/core/Lucid.kt
package org.afterlike.lucid.core

import best.azura.eventbus.core.EventBus
import net.minecraftforge.client.ClientCommandHandler
import org.afterlike.lucid.check.handler.CheckHandler
import org.afterlike.lucid.core.command.LucidCommand
import org.afterlike.lucid.core.handler.ConfigHandler
import org.afterlike.lucid.core.handler.DelayedTaskHandler
import org.afterlike.lucid.data.handler.impl.InteractionHandler
import org.afterlike.lucid.data.handler.impl.NetworkHandler
import org.afterlike.lucid.data.handler.impl.PlayerHandler
import org.afterlike.lucid.data.handler.impl.WorldHandler
import org.apache.logging.log4j.LogManager

class Lucid {

    private val logger = LogManager.getLogger(Lucid::class.java)

    companion object {
        @JvmStatic
        val INSTANCE: Lucid = Lucid()
    }

    val eventBus = EventBus()
    val checkHandler = CheckHandler

    fun initialize() {
        val start = System.nanoTime()
        logger.info("Initializing Lucid...")

        eventBus.subscribe(DelayedTaskHandler)
        eventBus.subscribe(CheckHandler)

        // data handlers
        eventBus.subscribe(PlayerHandler)
        eventBus.subscribe(NetworkHandler)
        eventBus.subscribe(WorldHandler)
        eventBus.subscribe(InteractionHandler)

        checkHandler.initialize()

        logger.info("Initialized Lucid in {}ms.", String.format("%.2f", (System.nanoTime() - start) / 1_000_000.0))
    }

    fun lateInitialize() {
        ClientCommandHandler.instance.registerCommand(LucidCommand())
        ConfigHandler.load()
    }
}
